crates/symbolicator-service/src/caching/fs.rs

```diff
@@ -103,39 +104,53 @@ impl Cache {
     pub(super) fn check_expiry(
         &self,
         path: &Path,
-    ) -> io::Result<(CacheContents<ByteView<'static>>, ExpirationTime)> {
-        // We use `mtime` to keep track of both "cache last used" and "cache created" depending on
-        // whether the file is a negative cache item or not, because literally every other
-        // filesystem attribute is unreliable.
-        //
-        // * creation time does not exist pre-Linux 4.11
-        // * most filesystems are mounted with noatime
-        //
+    ) -> io::Result<(CacheEntry<ByteView<'static>>, ExpirationTime)> {
+        let fs_metadata = path.metadata()?;
+        tracing::trace!("File `{}` length: {}", path.display(), fs_metadata.len());
+
+        // Open the metadata file if possible
+        let external_metadata: Option<Metadata> = self.read_metadata(path)?;
+
+        // Get mtime and ctime from the metadata file if possible. Otherwise, fall
+        // back to FS metadata.
+        let (atime, ctime) = match external_metadata.as_ref() {
+            Some(md) => (md.time_accessed, md.time_created),
+            None => {
+                // We use `mtime` to keep track of both "cache last used" and "cache created" depending on
+                // whether the file is a negative cache item or not, because literally every other
+                // filesystem attribute is unreliable.
+                //
+                // * creation time does not exist pre-Linux 4.11
+                // * most filesystems are mounted with noatime
+                let mtime = fs_metadata.modified()?;
```

Comment:
Only here we have to query the metadata, so I would move the `metadata()` call in here.

Reply:
I don't know how hot the function is, but it does save a syscall if you don't unconditionally check it, which may or may not matter.